### PR TITLE
DrivAerML: lr=7e-4 at 4L/320d (vs current lr=5e-4)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current DrivAerML best (5.027%) was set with the default lr=5e-4 at 4L/320d+T_max=30. For AirfRANS, lr=7e-4 proved to be the optimal learning rate for the 4L/256d architecture. DrivAerML may similarly benefit from a slightly higher learning rate — the 4L/320d architecture has more capacity, and a higher lr may help it explore the parameter space more effectively within the 180-minute timeout. This tests whether AirfRANS's optimal lr generalizes to the DrivAerML problem at the current best architecture.

## Instructions
Train on DrivAerML with 4L/320d architecture, lr=7e-4, T_max=30:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --num-layers 4 \
  --hidden-dim 320 \
  --use-fourier-features \
  --no-ema \
  --optimizer adamw \
  --lr 7e-4 \
  --weight-decay 1e-2 \
  --grad-clip 1.0 \
  --scheduler cosine \
  --T-max 30 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-4L320d-lr7e4
```

Key changes vs baseline:
- `--lr 7e-4` (baseline: 5e-4)
- Everything else identical to the current best DrivAerML configuration

IMPORTANT: DrivAerML requires `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`. Do not change these flags.

Report `val_primary/surface_rel_l2_pct` at end of training.

## Baseline
Current best metrics (DrivAerML):
- `val_primary/surface_rel_l2_pct`: **5.027%**
- Architecture: 4L/320d + Fourier, no-EMA, AdamW, lr=5e-4, WD=1e-2, gc=1.0, T_max=30
- Baseline PR: #2648 (zoro), 257 epochs, 180-min timeout
- Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --num-layers 4 --hidden-dim 320 --use-fourier-features --no-ema --optimizer adamw --lr 5e-4 --weight-decay 1e-2 --grad-clip 1.0 --scheduler cosine --T-max 30 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### DrivAerML: lr=7e-4 at 4L/320d — NEGATIVE (NaN divergence at epoch 113)

| Metric | Baseline (lr=5e-4) | This run (lr=7e-4) | Delta |
|--------|-------------------|-------------------|-------|
| `val_primary/surface_rel_l2_pct` | **5.027%** (ep 257) | **8.884%** (ep 107) | +76.7% worse |
| Epochs before divergence | 257+ (stable) | 113 (then NaN) | — |

**W&B run:** [`zdl5ug6y`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/zdl5ug6y)

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 7e-4 --weight-decay 1e-2 \
  --grad-clip 1.0 --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 --epochs 999 \
  --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 \
  --max-eval-batches 200 --agent askeladd \
  --wandb-name "askeladd/drivaerml-lr7e4-4L320d" \
  --wandb-group drivaerml-4L320d-lr7e4
```

**Peak memory:** Not captured (process killed after NaN divergence).

### What happened

**lr=7e-4 is too high for DrivAerML at 4L/320d — causes gradient explosion and NaN divergence.**

The training converged normally through epoch 112 (best 8.884% at ep107), showing steady progress:
- Cosine trough progression: ep15=25.2% → ep45=13.4% → ep75=14.3% → ep105=8.9%
- Grad norms were stable at 1.4-1.9

At epoch 113 (cosine LR near peak at lr=6.92e-04), the grad norm exploded from 1.77 to **347.48** — a 196x spike. Despite `--grad-clip 1.0`, this single catastrophic gradient event destroyed the model weights, producing NaN from epoch 114 onward.

**Key insight:** Even before divergence, lr=7e-4 was on a worse trajectory than the baseline lr=5e-4. At epoch 107 (best before crash), the metric was 8.884% — the baseline at a comparable stage was already lower. The higher lr may explore faster initially but the cosine peak oscillations at 7e-4 are too destructive for DrivAerML's loss landscape, especially at higher epoch counts where the model is near a basin and small perturbations can eject it.

**Why it worked for AirfRANS but not DrivAerML:**
- AirfRANS uses 4L/256d (smaller) and benefits from aggressive exploration of its phase-transition dynamics
- DrivAerML's 4L/320d has more parameters and the loss landscape appears smoother — it benefits from longer, more stable training rather than aggressive exploration
- DrivAerML's 394 batches/epoch creates higher effective gradient variance per cosine cycle

### Suggested follow-ups

1. **lr=6e-4:** Try a smaller bump from the baseline — may capture some of the faster initial convergence without the instability. The divergence threshold appears to be between 5e-4 and 7e-4.
2. **lr=7e-4 + grad-clip=0.5:** Tighter clipping might prevent the catastrophic gradient spike that killed training. The baseline uses gc=1.0 which wasn't enough.
3. **lr=7e-4 + warmup:** A warmup period could stabilize early training and allow the higher lr to work without the cosine peak destruction.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2787 because ADVISOR: lr=7e-4 at 4L/320d: 8.884% vs 4.619% baseline (92% worse), crashed at ep113 with grad norms spiking to 347. The 4L/320d architecture is also superseded by 4L/512d (current baseline). Good diagnosis on the grad norm dynamics. New assignment incoming: a cross-benchmark Lookahead ablation — a completely unexplored lever that could be impactful like the no-EMA finding..
